### PR TITLE
Update KWS, MSNoise, Signalmixer Data Loaders & Evaluation Notebook, Add New Scripts for Mixed Signals

### PR DESCRIPTION
**Pull request by EyubogluMerve**
## Updates 

- Updates on ***kws20.py***:
    - Dataset functions for signalmixer are added to the KWS data loader.
    - White Noise application is updated using SNR levels (generated between [-5, 20] dB).
- Updates on ***msnoise.py***:
    - Unnecessary reading operations are removed (**filter_classes** & **filter_dtype**)
    - A probability array is added as a parameter to MSnoise class. (Holds the noise type probabilities)
    - Naming differences in the raw Train/Test data are handled during download operation.
- Updates on ***signalmixer.py***:
    - Several parameters are implemented:
        ***apply_prob***: noise application probability (0.8, default)
        ***snr_range***: SNR range to be applied to the dataset ([-5 ,10] dB SNR, default)
        ***noise_type***: array that contains the noise type names (All Noise Types, default)

- Evaluation Notebook for KWS is updated.
- Scripts for KWS NAS & v3 models are added for mixing with MSnoise applications.

### CODE REVIEW

**Review by rotx-maxim — CHANGES_REQUESTED** [review]
This is missing the changes from #266, #269, and possibly #280.

File: datasets/msnoise.py [hunk 1]
```diff
@@ -228,7 +180,7 @@ def __getitem__(self, index):
         target = int(self.targets[rnd_num])
 
         if self.quantize:
-            inp /= 256
+            inp = inp / 256
```
Comment (alicangok) [inline datasets/msnoise.py:183]: Please convert this to the original version.

File: datasets/msnoise.py [hunk 2]
```diff
@@ -237,44 +189,53 @@ def __reshape_audio(self, audio, row_len=128):
...
+
+            if self.desired_probs is None or len(self.desired_probs) != len(labels):
+                self.desired_probs = []
+                print('Each class will be selected using the same probability!')
+                label_count = len(labels)
+                for i in range(label_count):
+                    self.desired_probs.append(1/label_count)
+
+            elif np.sum(self.desired_probs) != 1:
+                print('Sum of the probabilities is not 1!')
```
Comment (alicangok) [inline datasets/msnoise.py:225]: Also print a helpful note to the user, such as "Carrying out the proper normalization"

File: datasets/signalmixer.py [hunk 3]
```diff
@@ -28,109 +41,111 @@ class signalmixer:
...
+                    inp = self.single_snr_mixer(inp)
+        return inp.type(torch.FloatTensor), int(target)
 
     def __len__(self):
-        return len(self.mixed_signal)
+        return len(self.data)
 
-    def snr_mixer(self):
-        ''' creates mixed signal dataset using the SNR level and the noise dataset
+    def single_snr_mixer(self, signal):
```
Comment (alicangok) [inline datasets/signalmixer.py:94]: Consider renaming this function to simply "snr_mixer"

File: datasets/signalmixer.py [hunk 4]
```diff
@@ -28,109 +41,111 @@ class signalmixer:
...
-        clean = clean * scalarclean
+        signal = signal * scalarclean
 
-        scalarnoise = 1 / rms_noise.reshape(-1, 1, 1)
+        scalarnoise = 1 / rms_noise
         noise = noise * scalarnoise
 
-        cleanfactor = 10**(snr/20)
-        noisyspeech = cleanfactor * clean + noise
+        cleanfactor = 10**(snr_val/20)
```
Comment (alicangok) [inline datasets/signalmixer.py:116]: For better readability, consider keeping the `self.snr_range` variable (don't assign it to `snr`), and replacing `snr_val` with `snr`.

**Review by alicangok — CHANGES_REQUESTED** [review]
Could you also have a look at `scripts/evaluate_kws20_v3_SNR10dB.sh`, and check that the `MixedKWS20_10dB` dataset still works, with all the changes we made to the kws20 and msnoise datasets?

Even better, I think it would be best if we move the `MixedKWS20_10dB` dataset into `signalmixer.py` for backward compatibility purposes, then delete the unmaintained `mixedkws.py`. Having two files for the same purpose can be confusing.

**Comment by EyubogluMerve** [comment]
> Could you also have a look at `scripts/evaluate_kws20_v3_SNR10dB.sh`, and check that the `MixedKWS20_10dB` dataset still works, with all the changes we made to the kws20 and msnoise datasets?
> 
> Even better, I think it would be best if we integrate `MixedKWS20_10dB` dataset into `signalmixer.py` for backward compatibility purposes, then delete the unmaintained `mixedkws.py`. Having two files for the same purpose can be confusing.

I created the same dataset `(MixedKWS20_10dB)` using `signalmixer.py` data loader under `kws20.py` file. Also, I removed the `mixedkws.py`.  In addition, I made some fixes in `msnoise.py` & `kws20.py`. 

I checked all the evaluation and training scripts just in case. They're all working correctly.

**Review by rotx-maxim — CHANGES_REQUESTED** [review]
Please address https://github.com/MaximIntegratedAI/ai8x-training/pull/299#pullrequestreview-1934358167

File: notebooks/KWS_Noise_Evaluation.ipynb
Comment (alicangok) [inline notebooks/KWS_Noise_Evaluation.ipynb:1]: I would suggest the following replacement text for introducing signalmixer:

"The signalmixer data loader has several parameters:
 
- The `apply_prob` parameter determines the percentage of the samples on which additive noise will be applied (default setting: 0.8 (%80))
- The `snr_range` parameter modifies the additive noise power, such that the resulting signal falls between the specified SNR limits (default setting: [-5, 10] dB SNR)
- The `noise_type` parameter can be used provide a list of noise types to be applied (default setting is to apply all possible noise types in the dataset).

File: datasets/signalmixer.py [hunk 5]
```diff
@@ -28,109 +28,109 @@ class signalmixer:
...
-        snr = self.snr
+        if self.snr_range[0] == self.snr_range[-1]:
+            snr = self.snr_range[0]
+        else:
+            snr = np.random.uniform(self.snr_range[0], self.snr_range[-1])
+        snr = torch.tensor(snr)
 
-        rmsclean = torch.sqrt(torch.mean(clean.reshape(
-            clean.shape[0], -1)**2, 1, keepdims=True)).unsqueeze(1)
+        rmsclean = torch.sqrt(torch.mean(signal**2))
```
Comment (ermanok) [inline datasets/signalmixer.py:95]: Please check the variable names like follows:
rmsclean -> rms_clean
scalarclean -> scalar_clean

**Review by ermanok — COMMENTED** [review]
Minor suggestions for code refactor.

File: datasets/kws20.py [hunk 6]
```diff
@@ -111,6 +111,8 @@ def __init__(self, root, classes, d_type, t_type, transform=None, quantization_s
         if not self.benchmark:
             self.__filter_silence()
         else:
+            if len(classes) == 35:
```
Comment (rotx-maxim) [inline datasets/kws20.py:114]: Can we parametrize this instead of hardcoding a number of classes?

File: datasets/kws20.py [hunk 7]
```diff
@@ -885,6 +887,9 @@ def KWS_get_datasets(data, load_train=True, load_test=True, num_classes=6, bench
     if num_classes in (6, 11, 20):
         classes = next((e for _, e in enumerate(datasets)
                         if len(e['output']) - 1 == num_classes))['output'][:-1]
+    elif num_classes == 35:
```
Comment (rotx-maxim) [inline datasets/kws20.py:890]: see earlier comment

**Review by rotx-maxim — CHANGES_REQUESTED** [review]

**Review by ermanok — CHANGES_REQUESTED** [review]
Could you please add `--print-freq 100` to the training scripts? Now, the logs are too long to go over...

**Comment by alicangok** [comment]
@EyubogluMerve, could you please update your PR with the following patch I created?: [MervePR.patch](https://github.com/user-attachments/files/15873487/MervePR.patch)

List of changes:
- class names now follow the original dataset convention (e.g. `SILENCE` -> `_silence_`)
- training and validation examples for the `_silence_` class now come from all possible `_background_noise_` types, instead of exclusively using `running_tap` for validation (as we discussed in Teams)
- set `download=True` for the MSNoise dataset
- modifications and additions to comments and printed logs

---

Note: This PR contains the `train_kws12_nas_benchmark.sh` script, which requires a model checkpoint for the 12-class case which currently does not exist in the [ongoing PR to the synthesis repo](https://github.com/analogdevicesinc/ai8x-synthesis/pull/332). We will either need to remove this script from this PR, or add the trained checkpoint file to the synthesis repo. (If we decide on the latter, I will make a separate PR to the synthesis repo)

**Comment by alicangok** [comment]
@EyubogluMerve, could you apply [MervePR2.patch](https://github.com/user-attachments/files/15960033/MervePR2.patch) and update the PR?

This patch introduces constants to deal with the "data_type"s and replaces the `|` with the explicit `torch.logical_or()` for better readability.

**Review by rotx-maxim — APPROVED** [review]
